File: algorithms/0566.reshape_the_matrix/reshape_the_matrix.rs
impl Solution {
    pub fn matrix_reshape(mat: Vec<Vec<i32>>, r: i32, c: i32) -> Vec<Vec<i32>> {
        if (r * c) as usize == mat.len() * mat[0].len() {
            mat.concat()
                .chunks(c as usize)
                .map(|v| v.to_vec())
                .collect()
        } else {
            mat
        }
    }
}

File: algorithms/0566.reshape_the_matrix/reshape_the_matrix.cpp
class Solution {
public:
    vector<vector<int>> matrixReshape(vector<vector<int>>& mat, int r, int c) {
        if (mat.empty() || mat[0].size() * mat.size() != r * c) {
            return mat;
        }
        vector<vector<int>> result;
        vector<int> line;
        int count = 0;
        for (const auto&v: mat) {
            for (const int &n: v) {
                line.push_back(n);
                if (line.size() == c) {
                    result.push_back(line);
                    line.clear();
                }
            }
        }
        return result;
    }
};